linera-execution/src/wasm/wasmer.rs

```diff
@@ -273,27 +267,65 @@ impl From<wasmer::RuntimeError> for ExecutionError {
 }
 
 /// Serialized bytes of a compiled contract bytecode.
-///
-/// Each [`Module`] needs to be compiled with a separate [`Engine`] instance, otherwise Wasmer
-/// [panics](https://docs.rs/wasmer-middlewares/3.3.0/wasmer_middlewares/metering/struct.Metering.html#panic)
-/// because fuel metering is configured and used across different modules.
 pub struct CachedContractModule {
     compiled_bytecode: Bytes,
 }
 
+pub fn add_metering(bytecode: Bytecode) -> anyhow::Result<Bytecode> {
+    struct WasmtimeRules;
```

Comment:
Minor: I think this is mostly a style suggestion, but I would create a `WasmFuelMetering` (maybe even in the `common` module so we can also use it with Wasmtime) and implement `gas_metering::Rules` for it, as well as having a separate `impl` block with an `inject` associated function with the contents of this function starting at line 303.

I know this is subjective, but my reasoning is that:
- shorter functions are IMHO easier to read, so having the items it uses outside of it gives more focus to what it _does_
- it removes one level of indentation :laughing:

Reply:
I started out with a separate inline module and then realized that the function is pretty short even with the type inlined and it saves some imports :) I'm not too attached to it though, I can split it out if you prefer. I do generally prefer not to use module names like ‘common’, ‘utils’, ‘helper’, ‘stuff’ though.

Reply:
Oh, right! We don't have a `common` module anymore :tada: I agree with the usage of too generic names, I just thought we still had that "runtime independent" module and could reuse it :sweat_smile: . Since we don't, I think it might make more sense to create a new module, with a more specific name (`instrumentation`, `fuel`, `metering`?)

Reply:
I've made an issue to track the hoisting-up: https://github.com/linera-io/linera-protocol/issues/1967